torchqdynamics/mesolve.py

```diff
@@ -69,7 +88,62 @@ def forward_adjoint(self, t, dt, phi):
 
 class MERouchon2(MERouchon):
     def forward(self, t, dt, rho):
-        raise NotImplementedError
+        """Compute rho(t+dt) using a Rouchon method of order 2.
+
+        NOTE: For fast time-varying Hamiltonians, this method is not order 2 because the
+        second-order time derivative term is neglected. This term could be added in the
+        zero-th order Kraus operator if needed, as M0 += -0.5j * dt**2 * \dot{H}.
+        """
+        # non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # build time-dependent Kraus operators
+        M0 = self.I - 1j * dt * H_nh - 0.5 * dt**2 * H_nh @ H_nh
+        M1s = 0.5 * sqrt(dt) * (self.jump_ops @ M0 + M0 @ self.jump_ops)
+
+        # compute rho(t+dt)
+        rho_ = kraus_map(rho, M1s)
+        rho = kraus_map(rho, M0) + rho_ + 0.5 * kraus_map(rho_, M1s)
+        return rho / btrace(rho)
 
     def forward_adjoint(self, t, dt, phi):
         raise NotImplementedError
+
+
+def kraus_map(rho, *operators):
+    """Compute the application of a Kraus map on an input density matrix.
+
+    This is equivalent to `torch.sum(op @ rho @ op.adjoint())`.
+
+    Kraus operators are batched together such that only a single matrix operation is
+    called. The use of einsum yields better performances on large matrices, but may
+    cause a small overhead on smaller matrices (N <~ 50).
+    """
+    if any(op.ndim > 3 or op.ndim < 2 for op in operators):
+        raise ValueError(
+            'Kraus operators should be 2-D tensors, or 3-D tensors if batched.'
+        )
+
+    ops_batch = torch.cat(
+        tuple(op if op.ndim == 3 else op.unsqueeze(0) for op in operators)
+    )
+    return torch.einsum('kin,...nm,kjm->...ij', ops_batch, rho, ops_batch.conj())
+
+
+def btrace(rho):
+    """Compute the batched trace of a tensor over its last two dimensions, and return a
+    tensor of the same number of dimensions as rho."""
+    return torch.einsum('...ii', rho).real[..., None, None]
```

Comment:
Functions should always do what they say and nothing more, so we should either rename the method (`btrace_real`, `dmtrace`, ...) and edit the doc to say we just take the real part, or move the `.real` elsewhere to avoid possible misuse later.